source/ed/osm2ed.cpp

```diff
@@ -902,52 +900,66 @@ void PoiHouseNumberVisitor::fill_housenumber(const uint64_t osm_id,
     house_numbers.push_back(OSMHouseNumber(str_to_int(it_hn->second), lon, lat, candidate_way->way_ref));
 }
 
-void PoiHouseNumberVisitor::fill_poi(const u_int64_t osm_id, const CanalTP::Tags& tags,
-        const double lon, const double lat, OsmObjectType osm_relation_type) {
-    if (!parse_pois)
-        return;
-
-    std::string ref_tag = "";
-    for(const auto& tags_type: tags_types){
-        if(tags.find(tags_type) != tags.end()){
-            ref_tag = tags_type;
-            break;
+/**
+ * the first rule that matches OSM object decides POI-type
+ * match : OSM object has all OSM tags required by rule
+ */
+const RuleOsmTag2PoiType* PoiHouseNumberVisitor::get_applicable_poi_rule(const CanalTP::Tags& tags) {
+    for (const auto& osm_rule: poi_params.rules) {
+        bool is_valid_rule = true;
+        for (const auto& osm_rule_tag: osm_rule.osm_tag_filters) {
+            const auto it_poi_tag = tags.find(osm_rule_tag.first);
+            if (it_poi_tag == tags.end() || it_poi_tag->second != osm_rule_tag.second) {
+                is_valid_rule = false;
```

Comment:
ah there might have been a misunderstanding, it's the above for loop that can be rewrote with contains_if

here is a raw example (I've not compiled it)
```
 const auto not_valid = contains_if(osm_rule.osm_tag_filters, [&](const auto& osm_rule_tag) {
            const auto it_poi_tag = tags.find(osm_rule_tag.first);
            return it_poi_tag == tags.end() || it_poi_tag->second != osm_rule_tag.second);
});
```